ga4gh/exceptions.py

```diff
@@ -0,0 +1,179 @@
+"""
+Exceptions for the GA4GH server. Each exception that can occur in the server
+is given a unique error code that is derived from its name.
+"""
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import zlib
+
+import ga4gh.protocol as protocol
+
+
+def convertException(exception):
+    """
+    Converts the specified exception that is not a subclass of the
+    ServerException into a ServerError so that it can be correctly
+    serialised and communicated back to the client.
+    """
+    serverException = ServerError()
+    serverException.message = str(exception)
+    return serverException
+
+
+class ServerException(Exception):
+    """
+    Superclass of all exceptions that can occur in the GA4GH reference
+    server.
+    """
+
+    def getMessage(self):
+        """
+        Returns the message that we map into the GA4GH protocol message.
+        For many exceptions we can simply set the message statically
+        for a given class, and this message will be returned by
+        default. For more elaborate messages that must be contructed
+        at run time, we can override this method and use state held
+        in instance variables in the exception.
+        """
+        return self.message
+
+    @classmethod
+    def getErrorCode(cls):
+        """
+        Returns an integer that can be used to identif this class.
+        """
+        # We use the CRC32 digest of the class name as a unique code.
+        # We follow the recommendation of the Python docs to ensure
+        # that this value is signed 32 bit integer.
+        code = zlib.crc32(cls.__name__) & 0xffffffff
```

Comment:
This could still potentially have collisions.  CRC does not have strong uniqueness guarantees.